README
version:    2.3
website:    http://mij.oltrelinux.com/net/greylite/
contact:    mij <mij@bitchx.it>
license:    BSD revised (see http://www.opensource.org/licenses/bsd-license.php)

Refer to the website for any information.

Contents:
Changes             relevant software changes of each version
greylite.c          greylite's source file. Compile and install this one
logmsg.*            logging modules
greylite.8          greylite's man page. Install in your OS
greydb.sql          sql commands to initialize the greylite sqlite database
greylite-verified2cdb.sh    Script to selectively consolidate greylite's
                    verified addresses into cdb (tcprules)
cidr2firstlast.c    a trivial tool to convert address blocks from CIDR format
                    to first-last as used in tcprules. Useful for whitelisting
                    addresses expressed in SPF blocks (dig +short TXT domain).
Makefile.gnumake    Makefile for GNUMake -- rename to Makefile to build with
                    gnumake
Makefile.bsdmake    Makefile for BSDMake -- rename to Makefile to build with
                    bsdmake
dnsblenv.c          source for the dnsblenv module, greylite module for polling
                    DNS Black Lists
ucspi2socket.c      source for the ucspi2socket module, greylite module for
                    interfacing greylite with SMTP servers via TCP
greylite-patterns.conf.sample           example ruleset for suspicion


COMPILING

You can compile two more modules along with greylite: dnsblenv and
ucspi2socket.

1. If you use BSDmake instead of GNUmake, rename Makefile.bsdmake to Makefile
first

2. compile and install with
$ make <arguments> install

Arguments is a list of zero of more of the following:
WITH_GEOIP=yes
    To compile greylite with support for GeoIP suspicion rules. The GeoIP
    API and library must be installed on the system

WITH_UCSPI2SOCKET=yes
    To compile the "ucspi2socket" module along with greylite.

WITH_DNSBLENV=yes
    To compile the "dnsblenv" module along with greylite. This module requires
    the c-ares library on the system.

WITH_CARES14=yes
    The dnsblenv by default looks for c-ares version 1.5, whose API is
    incompatible with version 1.4 (still widespread). If c-ares 1.4 is present
    on the system use this argument.



Changes
* 2.3   Mar 2008
    - stricter control on compliance of commands from the clients
    - raise sqlite tryout, lower log priority for useless DB lock messages

* 2.2   Feb 2008
    - fix "354 go ahead (in reply to RCPT TO command)" bug when responding
      to non-suspicious clients at the 2nd delivery attempt

* 2.1   Jan 2008
    - STARTTLS now handled explicitly
    - logging made atomic, fixes possible log mixes with high concurrencies
    - example file for a tolerant suspicion ruleset included in the distribution

* 2.0   Jan 2008
    - greylite can now work as a generic SMTP proxy
    - command errors in client behaviour analysis
    - completely new command session control, now able to extract envelope
      information out of any SMTP command session even with
      unrecognized/custom commands. SMTP AUTH still passes transparently
      automatically

* 1.4   Dec 2007
    - greetdelay and client behaviour analysis
    - dnsblenv module for combining greylite's suspicion with DNS/RBL blacklists
    - verified servers are automatically pruned after 25 days of inactivity
    - suspicion supports inverted rules and more kinds: behaviour, envelope

* 1.3   Dec 2007
    - suspicion now supports a manifold of rule types: regex, envvar, geoip

* 1.2   Nov 2007
    - fix possible buffer overflow
    - fix collision handling on sqlite db
    - suspicion
    - earlier detection of "dead clients" (shorter hangs)
    - new man page
    - refined logging

* 1.1   Nov 2007
    - auto-pruning stale pending requests in the database
    - correctly handle "dead" clients, do not wait indefinitely anymore
    - insist on database locking up to five times, w/ exponential delays

* 1.0   Nov 2007
    - first public release


logmsg.c
#include "logmsg.h"

#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <sys/types.h>
#include <unistd.h>

#define LOGMSG_BUFLEN       1024

char logmsg_msgbuf[LOGMSG_BUFLEN];
int logmsg_inclpid;         /* true iff process ID has to be prefixed to log msgs */
pid_t logmsg_mypid;         /* if log_prefixpid, then set to the PID of the process */


int logmsg_init(const char *prefix) {
    char *x;

    logprefix = malloc(strlen(prefix)+1);
    strcpy(logprefix, prefix);

    x = getenv("LOGTHRESHOLD");
    if (x != NULL) {
        loglevel = (int)strtol(x, (char **)NULL, 10);
        if (loglevel < 0 || loglevel > 7) {
            loglevel = LOG_ERR;
        }
    } else
        loglevel = LOG_ERR;

    x = getenv("LOGPID");
    if (x != NULL) {
        logmsg_inclpid = 1;
        logmsg_mypid = getpid();
    } else {
        logmsg_inclpid = 0;
    }

    return 0;
}

int logmsg(int prio, char *fmt, ...) {
    va_list ap;

    if (prio > loglevel) return 0;

    /* include Process ID? */
    if (logmsg_inclpid) {
        snprintf(logmsg_msgbuf, LOGMSG_BUFLEN, "(%d) %s%s", logmsg_mypid, logprefix, fmt);
    } else {
        snprintf(logmsg_msgbuf, LOGMSG_BUFLEN, "%s%s", logprefix, fmt);
    }

    if (fmt[strlen(fmt)-1] != '\n') {
        strncat(logmsg_msgbuf, "\n", LOGMSG_BUFLEN);
    }

    va_start(ap, fmt);
    vfprintf(stderr, logmsg_msgbuf, ap);
    va_end(ap);

    return 0;
}



ucspi2socket.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <fcntl.h>
#include <sys/uio.h>
#include <unistd.h>
#include <poll.h>
#include <errno.h>

#include "logmsg.h"

/* file descriptors for reading from and writing to network */
#define NETWORK_READ_FD                 0
#define NETWORK_WRITE_FD                1

/* after being idle for these many secs the connection is terminated */
#define TIMEOUT                         40

/* no command lines will ever be longer than this many bytes (adjust as you need) */
#define BUFLEN                          1024

extern int loglevel;

int main(int argc, char *argv[]) {
    /* connect to upstream server */
    int srvsock;
    struct sockaddr_in srvsa;
    struct pollfd fds[2];
    int ret, len, cnt;
    char srvaddr[16], buf[BUFLEN];
    unsigned short int srvport;


    logmsg_init("ucspi2socket: ");

    /* sanity checks */
    if (argc < 2) {
        logmsg(LOG_CRIT, "Specify upstream server IP as command line argument.");
        exit(100);
    }
    loglevel = LOG_DEBUG;
    
    /* set the log level */
    if (getenv("LOGTHRESHOLD") != NULL) {
        char *x = getenv("LOGTHRESHOLD");
        sscanf(x, "%u", & loglevel);
    }

    /* extract the address of the upstream server to connect to */
    sscanf(argv[1], "%s", srvaddr);

    /* determine the port number to connect to */
    if (argc < 3 || sscanf(argv[2], "%hu", & srvport) != 1) {
        /* port parameter missing, default to 25 */
        srvport = 25;
    }
    logmsg(LOG_DEBUG, "Got server address '%s' port %hu", srvaddr, srvport);

    /* preparing to connect to the upstream server */
    srvsock = socket(AF_INET, SOCK_STREAM, 0);
    if (srvsock < 0) {
        logmsg(LOG_ERR, "Unable to create socket: %s.", strerror(errno));
        exit(100);
    }

    srvsa.sin_family = AF_INET;
    srvsa.sin_port = htons(srvport);
    srvsa.sin_addr.s_addr = inet_addr(srvaddr);
    memset(& srvsa.sin_zero, 0x00, sizeof(srvsa.sin_zero));

    /* connecting */
    ret = connect(srvsock, (struct sockaddr *) &srvsa, sizeof(srvsa));
    if (ret != 0) {
        logmsg(LOG_ERR, "Unable to connect socket: %s\n", strerror(errno));
        exit(100);
    }
    logmsg(LOG_DEBUG, "Connected successfully to %s:%hu, socket %d", srvaddr, srvport, srvsock); 

    /* preparing the sockets for non-blocking I/O */
    fcntl(NETWORK_READ_FD, F_SETFL, O_NONBLOCK);
    fcntl(srvsock, F_SETFL, O_NONBLOCK);

    /* preparing for poll() */
    fds[0].fd = NETWORK_READ_FD;
    fds[0].events = POLLIN;

    fds[1].fd = srvsock;
    fds[1].events = POLLIN;

    /* exchange data */
    while ((ret = poll(fds, 2, TIMEOUT*1000)) != -1) {
        /* data to read */
        if (fds[0].revents & POLLIN) {
            /* client ---> server */
            len = read(NETWORK_READ_FD, buf, BUFLEN);
            if (len <= 0) exit(0);
            cnt = 0;
            do {    /* beware of srvsock being O_NONBLOCK ... */
                ret = write(srvsock, buf+cnt, len-cnt);
                if (ret < 0) {
                    logmsg(LOG_ERR, "Unable to write %d bytes to the server: %s", strerror(errno));
                    exit(100);
                }
                cnt += ret;
            } while (cnt < len);
        } else if (fds[1].revents & POLLIN) {
            /* server ---> client */
            len = read(srvsock, buf, BUFLEN);
            if (len <= 0) exit(0);
            if (write(NETWORK_WRITE_FD, buf, len) != len) {
                logmsg(LOG_ERR, "Unable to write %d bytes to the client.");
                write(srvsock, "QUIT\r\n", strlen("QUIT\r\n"));
                exit(0);
            }
        } else {
            logmsg(LOG_NOTICE, "Timeout while proxying data.");
            exit(100);
        }
    }

    
    return 0;
}



cidr2firstlast.c
/**
 * Translates address blocks from CIDR format to first-last address format.
 *
 * Usage:
 * ./cidr2firstlast  and feed address block (like "192.168.0.0/23") in standard input
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>

int main() {
    uint32_t addrfrom, addrto, bitmask;
    char cidr[20];
    char *mask;
    unsigned short int intmask, i;

    while (fgets(cidr, 20, stdin) != NULL) {
        for (i = 0; cidr[i] != '/'; i++);
        cidr[i] = '\0';
        mask = & cidr[i+1];
        intmask = (unsigned short int)strtol(mask, (char **)NULL, 0);

        addrfrom = inet_addr(cidr);
        bitmask = htonl((0x00000001 << (32-intmask))-1);
        addrto = htonl(bitmask | addrfrom);

        printf("%s/%u   -   %s to ", cidr, intmask, cidr);
        printf("%d.%d.%d.%d\n", addrto >> 24, (addrto & 0x00FF0000) >> 16, (addrto & 0x0000FF00) >> 8, addrto & 0x000000FF);
    }

    return 0;
}




dnsblenv.c
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <errno.h>
#include <netdb.h>
#include <arpa/nameser.h>
#include <sys/time.h>

/* asynchronous DNS library */
#include <ares.h>

#include "logmsg.h"


/* delay between queries in milliseconds */
#define DELAY_QUERY         300
/* a check for DNS response is checked every this many millisecs */
#define STEP_PERIOD         20

/* maximum number of RBL lists that is possible to configure */
#define MAX_RBLS            20


/* whether the domain has been resolved against any list */
int found = 0;
unsigned int numresponses = 0;

char *clientIP, lookupname[256];

#ifdef CARES14
void dnsresultcallback(void *arg, int status, unsigned char *abuf, int alen);
#else
void dnsresultcallback(void *arg, int status, int timeouts, unsigned char *abuf, int alen);
#endif

int main(int argc, char *argv[]) {
    char *dnsbls;
    char *tok, *toktracer;
    char revIP[16];
    in_addr_t cliaddr;
    ares_channel dnschan;
    unsigned int cnt, numrequests = 0;
    fd_set readfds, writefds;
    int answers[MAX_RBLS];


    if (argc < 2) {
        fprintf(stderr, "Specify the child to run as argument.\n");
        return 110;
    }

    logmsg_init("dnsblenv: ");

    dnsbls = getenv("DNSBL");
    if (dnsbls == NULL) {
        /* pass-through */
        execvp(argv[1], argv+1);
        logmsg(LOG_CRIT, "Unable to exec: %s\n", strerror(errno));
        return 110;
    }

    clientIP = getenv("TCPREMOTEIP");
    if (clientIP == NULL) {
        logmsg(LOG_CRIT, "Unable to fetch remote client address. Terminating.\n");
        return 110;
    }

    /* build hostname to look for */
    cliaddr = inet_addr(clientIP);
    sprintf(revIP, "%d.%d.%d.%d", (cliaddr & 0xFF000000) >> 24, (cliaddr & 0x00FF0000) >> 16, (cliaddr & 0x0000FF00) >> 8, cliaddr & 0x000000FF);

    if (ares_init(& dnschan) != ARES_SUCCESS) {
        logmsg(LOG_ERR, "Unable to open dns channel.\n");
        return 110;
    }

    FD_ZERO(&readfds);
    FD_ZERO(&writefds);

    memset(answers, 0x00, MAX_RBLS);

    toktracer = dnsbls;
    tok = strsep(& toktracer, " ");
    while (tok != NULL && !found && numrequests < MAX_RBLS) {
        numrequests++;

        sprintf(lookupname, "%s.%s", revIP, tok);
        
        logmsg(LOG_DEBUG, "Querying for %s\n", lookupname);

        /* perform DNS lookup */
        ares_query(dnschan, lookupname, ns_c_in, ns_t_a, dnsresultcallback, & answers[numrequests-1]);

        tok = strsep(& toktracer, " ");

        ares_fds(dnschan, &readfds, & writefds);

        /* insert a small delay between queries but poll the status meanwhile */
        cnt = 0;
        do {
            usleep(STEP_PERIOD*1000);
            cnt += STEP_PERIOD;
            ares_process(dnschan, & readfds, & writefds);
        } while (!answers[numrequests-1] && cnt < DELAY_QUERY);
    }
    if (numrequests == MAX_RBLS) {
        logmsg(LOG_CRIT, "Warning! Number of RBLs configured exceeds hardcoded limit (%u)", MAX_RBLS);
        execv(argv[1], argv+1);
        logmsg(LOG_CRIT, "Unable to run %s: %s.\n", argv[1], strerror(errno));
        exit(100);
    }

    /* cleanup uncompleted queries, if any */
    while (numrequests > numresponses) {
        /* insert a small delay between queries */
        usleep(STEP_PERIOD*1000);
        
        ares_fds(dnschan, &readfds, & writefds);
        ares_process(dnschan, & readfds, & writefds);
    }

    ares_destroy(dnschan);
    if (found) {
        char *varname;

        varname = getenv("DNSBL_VARNAME");
        if (varname == NULL) {
            varname = "BLACKLISTED";
        }
        setenv(varname, "", 1);
    }

    execv(argv[1], argv+1);
    logmsg(LOG_CRIT, "Unable to run %s: %s.\n", argv[1], strerror(errno));


    return 0;
}


/* arg is the domain of the list of the lookup */
#ifdef CARES14
void dnsresultcallback(void *arg, int status, unsigned char *abuf, int alen) {
#else
void dnsresultcallback(void *arg, int status, int timeouts, unsigned char *abuf, int alen) {
#endif
    struct hostent *he;
    char *name;
    int i;
    struct in_addr addr;

    /* keep the count of how many SERVERS responded */
    if (*(int *)arg == 0) numresponses++;

    /* increase number of responses for the list answering */
    *(int *)arg += 1;

    switch (status) {
        case ARES_SUCCESS:
            /* address resolved */
            found = 1;
#ifdef CARES14
            if (ares_parse_a_reply(abuf, alen, & he) != ARES_SUCCESS) {
#else
            if (ares_parse_a_reply(abuf, alen, & he, NULL, NULL) != ARES_SUCCESS) {
#endif
                logmsg(LOG_NOTICE, "Unable to parse DNS response\n");
                return;
            }
            for (i = 0, name = he->h_addr_list[0]; name != NULL; i++, name = he->h_addr_list[i]) {
                memcpy(&addr, name, he->h_length);
                logmsg(LOG_INFO, "'%s' listed as '%s'.\n", clientIP, inet_ntoa(addr));
            }
            ares_free_hostent(he);
            break;
        case ARES_ENODATA:
        case ARES_ENOTFOUND:
            /* domain does not exist */
            break;
        default:
            logmsg(LOG_INFO, "Generic error when querying '%s'.\n", (char *)arg);
    }
}

